[on_call_schedule] Add on-call schedule resource

daniel-m-dd opened the pull request:
Create new resource `datadog_on_call` to manage schedules used for on-call
https://docs.datadoghq.com/api/latest/on-call/#create-on-call-schedule

A schedule can have "empty member"; gap/void users, I would have liked to express it as `member{}` but this does not seem supported by old terraform version [see [here](https://github.com/hashicorp/terraform/pull/32463)]; ~the current workaround is `member{ user{} }` but there might be a better API here.~ ~Edit 1: updated to be `member {}` or `member {user_id=null}}`~
~Edit 2: updated to be `member_ids=[id, null,...]`~
Edit 3: updated to be `users=[id, null, ...]`

daniel-m-dd committed:
[on_call_schedule] Add on-call schedule resource

daniel-m-dd commented on datadog/fwprovider/resource_datadog_on_call_schedule.go:
```diff
@@ -0,0 +1,803 @@
+	return t.Format(time.RFC3339)
+}
+
+func (r *onCallScheduleResource) buildOnCallScheduleRequestBody(ctx context.Context, state *onCallScheduleModel) (*datadogV2.ScheduleCreateRequest, diag.Diagnostics) {
```
a lot of code is duplicated between Create and Update because the golang sdk use different go types for both actions. This might be simplified by potentially sharing some types in between the 2 operations. (Types are not the same as some minor differences exsit between the 2 operations, for example an PUT/Update might include a layer.id field; a POST/Create can't.

daniel-m-dd committed:
Generate doc

daniel-m-dd commented on datadog/fwprovider/resource_datadog_on_call_schedule.go:
```diff
@@ -0,0 +1,803 @@
+								},
+							},
+						},
+						"interval": schema.SingleNestedBlock{
```
we might want to create a special type here to express that at least one of seconds or days should be > 0
we can start with a validator

daniel-m-dd committed:
Fix empty members

daniel-m-dd committed:
Remove redundnat oncall schedule tests

daniel-m-dd committed:
on-call_schedules: Add support for empty members in on-call schedules

Bnjis committed:
update descriptions

bndbsh commented on datadog/fwprovider/resource_datadog_on_call_schedule.go:
```diff
@@ -0,0 +1,802 @@
+								"seconds": schema.Int64Attribute{
+									Optional:    true,
+									Computed:    true,
+									Description: "Extra seconds that may be added to extend the rotation beyond whole days.",
```
```suggestion
									Description: "For intervals that are not expressible in whole days, this will be added to `days`.",
```

From the doc it's not immediately clear that interval can also be less than one day

daniel-m-dd committed:
Update datadog/fwprovider/resource_datadog_on_call_schedule.go

Co-authored-by: Belhorma Bendebiche <amro@bndb.sh>

daniel-m-dd committed:
Refactor on-call schedule resource

- Replace the attributes/blocks 'team' and 'layer' attributes in favor of 'team_ids' and 'layer'.
- Add missing description
- Rename members to member and flatten the user block into a `user_id` attribute within member
- Deleted unused mcp.json configuration file.

daniel-m-dd committed:
Merge branch 'master' of github.com:DataDog/terraform-provider-datadog into on-call/schedules

daniel-m-dd committed:
Update doc for oncall schedule

daniel-m-dd committed:
Merge branch 'master' of github.com:DataDog/terraform-provider-datadog into on-call/schedules

daniel-m-dd committed:
Refactor on-call schedule resource to use member_ids instead of nested member blocks

bndbsh committed:
remove _id suffix from attributes and update readme

bndbsh committed:
fix example

bndbsh committed:
linter fixes

bndbsh committed:
update docs

bndbsh committed:
remove tags

daniel-m-dd committed:
Fix empty oncall teams

daniel-m-dd committed:
Remove tags from tests

bndbsh committed:
record cassettes

bndbsh commented:
/merge
